File: SortList.cc
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x):val(x), next(NULL) {}
};

void merge(ListNode* l1,  ListNode* l2, ListNode*& head, ListNode*& tail)
{
    ListNode dummy(-1);
    dummy.next = head;
    head = &dummy;
    while (l1 && l2) {
        if (l1->val <= l2->val) {
            head->next = l1;
            l1 = l1->next;
        } else {
            head->next = l2;
            l2 = l2->next;
        }
        head = head->next;
    }
    tail = l1 ? l1 : l2;
    head->next = tail;
    while (tail->next)
        tail = tail->next;
    head = dummy.next;
    tail->next = NULL;
}

ListNode *sortList(ListNode *head) {
    int k = 1;
    ListNode dummy(-1);
    dummy.next = head;

    while (1) {
        ListNode* prev = &dummy;
        ListNode* l1 = dummy.next;
        ListNode* l2 = dummy.next;
        while (1) {
            int i = 0;
            for (l2 = l1; (i < k) && l2; ++i ) 
            {
                if (i == k-1) {
                    ListNode* tmp = l2;
                    l2 = l2->next;
                    tmp->next = NULL;
                } else
                    l2 = l2->next;
            }   

            if (l2) {
                ListNode* mergedHead = NULL;
                ListNode* mergedTail = NULL;
                ListNode* post = l2;
                int j = 0;
                for (; (j < k) && post; ++j)
                {
                    if (j == k-1) {
                        ListNode* tmp = post;
                        post = post->next;
                        tmp->next = NULL;
                    } else 
                         post = post->next;
                }

                merge(l1, l2, mergedHead, mergedTail);
                prev->next = mergedHead;
                mergedTail->next = post;

                l1 = post;
                prev = mergedTail;
            } else
                break;
        }
        k *= 2;
        if (l1 == dummy.next)
            break;
    }
    return dummy.next;
}

int main(int argc, char** argv)
{
    ListNode* test = new ListNode(5);
    test->next = new ListNode(4);
    test->next->next = new ListNode(3);
    test->next->next->next = new ListNode(8);
    ListNode* res = sortList(test);
    while (res) {
        cout << res->val << ", ";
        res = res->next;
    }
    cout << endl;
    return 0;

}

File: SortList.java


 
public class SortList {
	 public static class ListNode {
	     int val;
	     ListNode next;
	     ListNode(int x) { val = x; }
	 }    
    private ListNode[] mergeTwoLists(ListNode head1, ListNode head2) {
        ListNode dummy = new ListNode(-1);
        ListNode newTail = dummy;
        while (head1 != null && head2 != null) {
            if (head1.val <= head2.val) {
                newTail.next = head1;
                head1 = head1.next;
                newTail = newTail.next;
            } else {
                newTail.next = head2;
                head2 = head2.next;
                newTail = newTail.next;
            }
        }
        newTail.next = (head1 == null) ? head2 : head1;
        for (; newTail.next != null; newTail = newTail.next);
        return new ListNode[]{dummy.next, newTail};
    }
    
    //Bottom-up merge sort. O(nlogn) time and O(1) space
    public ListNode sortList(ListNode head) {
        int k = 1;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        while (true) {
            ListNode preTail = dummy;
            ListNode head1 = preTail.next;
            for (; ;) {
                ListNode tail1 = head1;
                for (int count = 0; count < k - 1 && tail1 != null; tail1 = tail1.next, ++count);
                if (tail1 == null) {
                    break;
                }
                ListNode head2 = tail1.next, tail2 = head2;
                tail1.next = null;
                for (int count = 0; count < k - 1 && tail2 != null; tail2 = tail2.next, ++count);
                ListNode nextHead1 = null;
                if (tail2 != null) {
                    nextHead1 = tail2.next;
                    tail2.next = null;
                }
                ListNode[] mergedNodes = mergeTwoLists(head1, head2);
                preTail.next = mergedNodes[0];
                preTail = mergedNodes[1];
                head1 = nextHead1;
                preTail.next = head1;
            }
            if (head1 == dummy.next) {
                break;
            }
            k = k << 1;
        }
        return dummy.next;
    }

    private static ListNode createLinkedListFromArray(int[] nums) {
    	ListNode dummy = new ListNode(-1);
    	ListNode newTail = dummy;
    	for (int num : nums) {
    		newTail.next = new ListNode(num);
    		newTail = newTail.next;
    	}
    	return dummy.next;
    }

    private static void printListNodes(ListNode node) {
    	for (ListNode curNode = node; curNode != null; curNode = curNode.next) {
    		System.out.print(curNode.val + ", ");
    	}
    	System.out.println("");
    }

    public static void main(String[] args) {
    	int[] test = {4, 19, 14, 5, -3, 1, 8, 5, 11, 15};
    	ListNode head = createLinkedListFromArray(test);
    	System.out.println("Input: ");
    	printListNodes(head);
    	SortList sl = new SortList();
    	ListNode newHead = sl.sortList(head);
    	printListNodes(newHead);
    }
}